test/test_modules.py

```diff
@@ -206,6 +207,121 @@ def test_check_inplace(self, device, dtype, module_info):
             self.assertEqual(input_args[0].grad, input_arg_copy[0].grad)
 
 
+    @modules(module_db)
+    def test_non_contiguous_tensors(self, device, dtype, module_info):
+        # Check modules work with non-contiguous tensors
+
+        module_cls = module_info.module_cls
+        module_inputs = module_info.module_inputs_func(module_info, device=device, dtype=dtype,
+                                                       requires_grad=True)
+
+        # gradients needs to be retained to check for grad. This is useful when
+        # non-leafs are present in the graph.
+        def _retain_grad(item):
+            if isinstance(item, dict):
+                for i in item.values():
+                    _retain_grad(i)
+            elif isinstance(item, (tuple, list)):
+                for i in item:
+                    _retain_grad(i)
+            elif isinstance(item, torch.Tensor) and item.requires_grad:
+                item.retain_grad()
+
+        def _get_grads(obj):
+            if isinstance(obj, tuple):
+                return tuple(_get_grads(o) for o in obj)
+            elif isinstance(obj, dict):
+                return {name: _get_grads(o) for name, o in obj.items()}
+            elif isinstance(obj, torch.Tensor) and obj.requires_grad:
+                return obj.grad
+
+        def _zero_grad(obj):
+            if isinstance(obj, (tuple, list)) or isgenerator(obj):
+                for o in obj:
+                    _zero_grad(o)
+            elif isinstance(obj, dict):
+                for o in obj.values():
+                    _zero_grad(o)
+            elif isinstance(obj, (torch.Tensor, torch.nn.Parameter)) and obj.grad is not None:
+                obj.grad.detach_()
+                obj.grad.zero_()
```

Comment:
When we start to combine tests together, this will be very common operation to reset the gradients.